clarity/src/vm/analysis/type_checker/tests/mod.rs

```diff
@@ -3039,3 +3053,131 @@ fn test_principal_construct() {
     }
 }
+
+#[test]
+fn test_trait_args() {
+    let good = [
+        "(define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (call-foo (f <trait-foo>))
+            (contract-call? f foo)
+        )
+        (define-public (call-foo-outer (f <trait-foo>))
+            (begin
+                (call-foo f)
+            )
+        )",
+        "(define-trait trait-foobar
+            (
+                (foo () (response uint uint))
+                (bar () (response uint uint))
+            )
+        )
+        (define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (call-foo (f <trait-foo>))
+            (contract-call? f foo)
+        )
+        (define-public (call-foo-foobar (f <trait-foobar>))
+            (begin
+                (call-foo f)
+            )
+        )",
+    ];
+
+    let bad = ["(define-trait trait-bar
+            (
+                (bar () (response uint uint))
+            )
+        )
+        (define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (call-foo (f <trait-foo>))
+            (contract-call? f foo)
+        )
+        (define-public (call-foo-foobar (f <trait-bar>))
+            (begin
+                (call-foo f)
+            )
+        )"];
+
+    let contract_identifier = QualifiedContractIdentifier::transient();
+    let bad_expected = [CheckErrors::IncompatibleTrait(
+        TraitIdentifier {
+            name: ClarityName::from("trait-foo"),
+            contract_identifier: contract_identifier.clone(),
+        },
+        TraitIdentifier {
+            name: ClarityName::from("trait-bar"),
+            contract_identifier: contract_identifier.clone(),
+        },
+    )];
+
+    for good_test in good.iter() {
+        assert!(mem_type_check(&good_test).is_ok());
+    }
+
+    for (bad_test, expected) in bad.iter().zip(bad_expected.iter()) {
+        assert_eq!(expected, &mem_type_check(&bad_test).unwrap_err().err);
+    }
+}
+
+#[test]
+fn test_wrapped_trait() {
+    let good = [
+        "(define-trait trait-foo ((foo () (response uint uint))))
+        (define-public (call-foo-if (opt (optional <trait-foo>)))
+            (match opt
+                f (contract-call? f foo)
+                (ok u1)
+            )
+        )",
+        "(define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (call-foo (f <trait-foo>))
+            (unwrap! (contract-call? f foo) u2)
+        )
+        (define-public (call-foo-list (l (list 5 <trait-foo>)))
+            (ok (map call-foo l))
+        )",
+        "(define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (return-f (f <trait-foo>))
+            (if true (ok f) (err u1))
+        )
+        (define-public (call-foo (f <trait-foo>))
+            (match (return-f f)
+                f-prime (contract-call? f-prime foo)
+                e (err u1)
+            )
+        )",
```

Comment:
Can you also add checks for `(response <trait-foo> uint)` and `(response uint <trait-foo>)`?

Reply:
In the last example, `return-f` returns a `(response <trait-foo> uint)`. I can add a similar test for `(response uint <trait-foo>)`.

Reply:
Added in 3827fe371